src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildTriggerListener.java

```diff
@@ -82,4 +87,38 @@ public void onDeleted(final Run<?,?> run) {
             Timer.get().submit(() -> trigger.context.onFailure(new AbortException(run.getFullDisplayName() + " was deleted")));
         }
     }
+
+    private void updateDownstreamBuildAction(Run<?, ?> run) {
+        for (Cause cause : run.getCauses()) {
+            if (cause instanceof BuildUpstreamCause) {
+                BuildUpstreamCause buildUpstreamCause = (BuildUpstreamCause) cause;
+                Run<?, ?> upstream = buildUpstreamCause.getUpstreamRun();
+                if (upstream instanceof FlowExecutionOwner.Executable) {
+                    FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) upstream).asFlowExecutionOwner();
+                    if (owner == null) {
+                        LOGGER.log(Level.FINE, () -> "Unable to update DownstreamBuildAction for " + upstream + " node " + buildUpstreamCause.getNodeId());
+                        continue;
+                    }
+                    try {
+                        FlowExecution execution = owner.get();
+                        FlowNode node = execution.getNode(buildUpstreamCause.getNodeId());
+                        if (node == null) {
+                            LOGGER.log(Level.FINE, () -> "Unable to update DownstreamBuildAction for " + upstream + " node " + buildUpstreamCause.getNodeId());
+                            continue;
+                        }
+                        DownstreamBuildAction downstreamAction = node.getPersistentAction(DownstreamBuildAction.class);
+                        if (downstreamAction == null) {
+                            // Should only happen for builds already in the queue when this plugin is updated to include DownstreamBuildAction.
+                            downstreamAction = new DownstreamBuildAction(run.getParent());
+                            node.addAction(downstreamAction);
+                        }
+                        downstreamAction.setBuild(run);
+                        run.save();
```

Comment:
During my tests it seems that restarting the controller reset the `DownstreamAction` (at least its `buildNumber` property).
Do we need to save the action?

Reply:
`run.save()` ought to do that. If you suspect this is broken somehow, it should be possible to reproduce using `JenkinsSessionRule`: run through this scenario in one session, then restart Jenkins and recheck the metadata.

Reply:
Oh, I bet the problem is that the `FlowNodeStorage` have optimizations to try to only save nodes that have changed, and setting a field on a single action by itself will not mark the `FlowNode` as updated in a way that the storages will recognize. I will look into it.

Reply:
Mutable fields :facepalm: maybe better to just make fields `final` and use `replaceAction`?

Reply:
Yes that was my plan